### PR TITLE
[Fuchsia][cmake] Avoid referencing cxx_shared in compiler-rt

### PR DESCRIPTION
**Pull request by ilovepi**
After https://github.com/llvm/llvm-project/pull/80007 Fuchsia builds are
now always building cxx_shared for arm64 and x64 Linux. Ultimately, this
is because the LIBCXX_ENABLE_SHARED is not used in compiler-rt to select
the correct libc++ target, and because cxx_shared is now always defined,
it is selected as a dependency when building runtimes tests.

### CODE REVIEW

File: compiler-rt/test/fuzzer/CMakeLists.txt [hunk 1]
```diff
@@ -73,11 +73,10 @@ endmacro()
 
 test_fuzzer("default")
 if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
-  if(TARGET cxx_shared)
-    test_fuzzer("libc++" DEPS cxx_shared)
-  endif()
-  if(TARGET cxx_static)
+  if(LIBCXX_ENABLE_STATIC)
     test_fuzzer("static-libc++" DEPS cxx_static)
+  elseif(LIBCXX_ENABLE_SHARED)
```
Comment (petrhosek) [inline compiler-rt/test/fuzzer/CMakeLists.txt:78]: The previous version test both static and shared if both were enabled, I think we should preserve that.

```suggestion
  if(LIBCXX_ENABLE_SHARED)
```

**Review by petrhosek — APPROVED** [review]

**Review by ldionne — COMMENTED** [review]
Does someone understand how the layering works for compiler-rt? Libc++ links against compiler-rt itself, so how can we depend on libc++ from compiler-rt?

**Comment by ilovepi** [comment]
> Does someone understand how the layering works for compiler-rt? Libc++ links against compiler-rt itself, so how can we depend on libc++ from compiler-rt?

Aren't the builtins the only dependency? IIRC builtins are special in the runtimes build, and is handled that way to break the cyclic dependency in the build graph. @petrhosek can speak more to that though.

**Comment by ldionne** [comment]
Thanks for the explanation!

> That's correct, we build builtins separately by using https://github.com/llvm/llvm-project/tree/main/compiler-rt/lib/builtins as the top-level CMake build directory

Would it make sense to separate this out of `compiler-rt` then? This sounds like a slightly brittle setup dependency-wise.

> libc++ is currently only used by libFuzzer, ORC and XRay plus all unit tests (since it's a dependency of GoogleTest)

Is there a reason why libFuzzer, ORC and XRay need to explicitly mention the just-built libc++? Couldn't they use whatever standard library is provided by the compiler, which in the case of a bootstrapping build would be the just-built libc++? That would be cleaner than having subparts of compiler-rt depending directly on libc++ targets.

> the other runtimes in compiler-rt like sanitizers only depend on libc++abi

Do you know what they depend on libc++abi for?

**Pull request merged**